pubsub/google/cloud/pubsub_v1/publisher/batch/base.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2017, Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import abc
+import collections
+
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseBatch(object):
+    """The base batching class for Pub/Sub publishing.
+
+    Although the :class:`~.pubsub_v1.publisher.batch.thread.Batch` class, based
+    on :class:`threading.Thread`, is fine for most cases, advanced
+    users may need to implement something based on a different concurrency
+    model.
+
+    This class defines the interface for the Batch implementation;
+    subclasses may be passed as the ``batch_class`` argument to
+    :class:`~.pubsub_v1.client.PublisherClient`.
+    """
+    def __len__(self):
+        """Return the number of messages currently in the batch."""
+        return len(self.messages)
+
+    @property
+    @abc.abstractmethod
+    def client(self):
```

Comment:
Having the client on a concurrency-safe object will not work, it is not safe to pass the client across the concurrency boundary. It *is* okay for the concrete thread-based batch class to reference the client, however.